components/playground/main.go

```diff
@@ -126,7 +130,66 @@ Examples:
 			}
 			environment.SetGlobalEnv(env)
 
-			return p.bootCluster(env, opt)
+			var booted uint32
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			go func() {
+				sc := make(chan os.Signal, 1)
+				signal.Notify(sc,
+					syscall.SIGHUP,
+					syscall.SIGINT,
+					syscall.SIGTERM,
+					syscall.SIGQUIT,
+				)
+
+				sig := (<-sc).(syscall.Signal)
+				atomic.StoreInt32(&p.curSig, int32(sig))
+				fmt.Println("Playground receive signal: ", sig)
+
+				// if bootCluster is not done we just cancel context to make it
+				// clean up and return ASAP and exit directly after timeout.
+				// Note now bootCluster can not learn the context is done and return quickly now
+				// like while it's downloading component.
+				if atomic.LoadUint32(&booted) == 0 {
+					cancel()
+					time.AfterFunc(time.Second, func() {
+						os.Exit(0)
+					})
+					return
+				}
+
+				go p.terminate(sig)
+				// If user try double ctrl+c, force quit
+				sig = (<-sc).(syscall.Signal)
+				atomic.StoreInt32(&p.curSig, int32(syscall.SIGKILL))
+				if sig == syscall.SIGINT {
+					p.terminate(syscall.SIGKILL)
+				}
+			}()
+
+			// TODO: we can set Pdeathsig of Cmd.SysProcAttr(linux only) in all the Cmd we started to kill
```

Comment:
Can we implement this in the current PR?

Reply:
#719 #720 
I will try but not in this PR.
Keeping this pr small and do one thing.